exchange/bitswap/wantmanager.go

```diff
@@ -137,36 +137,49 @@ func (mq *msgQueue) runQueue(ctx context.Context) {
 	for {
 		select {
 		case <-mq.work: // there is work to be done
-
-			err := mq.network.ConnectTo(ctx, mq.p)
-			if err != nil {
-				log.Noticef("cant connect to peer %s: %s", mq.p, err)
-				// TODO: cant connect, what now?
-				continue
-			}
-
-			// grab outgoing message
-			mq.outlk.Lock()
-			wlm := mq.out
-			if wlm == nil || wlm.Empty() {
-				mq.outlk.Unlock()
-				continue
-			}
-			mq.out = nil
-			mq.outlk.Unlock()
-
-			// send wantlist updates
-			err = mq.network.SendMessage(ctx, mq.p, wlm)
-			if err != nil {
-				log.Noticef("bitswap send error: %s", err)
-				// TODO: what do we do if this fails?
-			}
+			mq.doWork(ctx)
 		case <-mq.done:
 			return
 		}
 	}
 }
 
+func (mq *msgQueue) doWork(ctx context.Context) {
+	// allow a minute for connections
+	// this includes looking them up in the dht
+	// dialing them, and handshaking
+	conctx, cancel := context.WithTimeout(ctx, time.Minute)
+	defer cancel()
```

Comment:
> Do we need explicit sub-part timeouts at all?  Can't the caller
> specify their own timeout when they submit a task?

we could if all the callers/roots had timeouts enforced somehow. we dont do this perfectly yet.

> But maybe the doWork context
> should be a union of the runQueue context and a second context that
> gets passed in via the channel?  

could use https://github.com/jbenet/go-context/blob/master/ext/dagctx.go for this.